arm/curve25519/curve25519_x25519.S

```diff
@@ -26,841 +38,1317 @@
 
-        mov     i, #253
+        mov     x0, #(1<<26)-1
+        mov     v30.d[0], x0
+        mov     v30.d[1], x0
 
-curve25519_x25519_scalarloop:
+        mov     x0, #0x07fffffe07fffffe
+        sub     x1, x0, #0xfe-0xb4
+        sub     x0, x0, #2
 
-// sm = xm + zm; sn = xn + zn; dm = xm - zm; dn = xn - zn
+        stp     x0, x1, [mask1]
+        ldp     d29, d28, [mask1]
 
-        sub_twice4(dm,xm,zm)
-        add_twice4(sn,xn,zn)
-        sub_twice4(dn,xn,zn)
-        add_twice4(sm,xm,zm)
+// The main loop over (modified) bits from i = 254, ..., i = 0 (inclusive);
+// we explicitly skip bit 255 because it should be forced to zero initially.
+// This is a classic Montgomery ladder using a "swap" variable.
+// It's assumed x0 = i at the start of the loop, but that is volatile and
+// needs to be reloaded from memory at the end of the loop.
 
-// ADDING: dmsn = dm * sn
-// DOUBLING: mux d = xt - zt and s = xt + zt for appropriate choice of (xt,zt)
+        str     xzr, [swap]
+        mov     x0, #254
+        str     x0, [i]
 
-        mul_4(dmsn,sn,dm)
+curve25519_x25519_scalarloop:
 
-        lsr     x0, i, #6
-        ldr     x2, [sp, x0, lsl #3]    // Exploiting scalar = sp exactly
-        lsr     x2, x2, i
+        lsr     x1, x0, #6
+        ldr     x2, [sp, x1, lsl #3]    // Exploiting scalar = sp exactly
+        lsr     x2, x2, x0
         and     x2, x2, #1
 
-        cmp     swap, x2
-        mov     swap, x2
-
-        mux_4(d,dm,dn)
-        mux_4(s,sm,sn)
-
-// ADDING: dnsm = sm * dn
-
-        mul_4(dnsm,sm,dn)
-
-// DOUBLING: d = (xt - zt)^2
-
-        sqr_4(d,d)
-
-// ADDING: dpro = (dmsn - dnsm)^2, spro = (dmsn + dnsm)^2
-// DOUBLING: s = (xt + zt)^2
-
-        sub_twice4(dpro,dmsn,dnsm)
-        sqr_4(s,s)
-        add_twice4(spro,dmsn,dnsm)
-        sqr_4(dpro,dpro)
-
-// DOUBLING: p = 4 * xt * zt = s - d
-
-        sub_twice4(p,s,d)
-
-// ADDING: xm' = (dmsn + dnsm)^2
-
-        sqr_4(xm,spro)
-
-// DOUBLING: e = 121666 * p + d
-
-        mov     x1, 0xdb42
-        orr     x1, x1, 0x10000
-        cmadd_4(e,p,d)
-
-// DOUBLING: xn' = (xt + zt)^2 * (xt - zt)^2 = s * d
-
-        mul_4(xn,s,d)
-
-// ADDING: zm' = x * (dmsn - dnsm)^2
-
-        mul_4(zm,dpro,pointx)
-
-// DOUBLING: zn' = (4 * xt * zt) * ((xt - zt)^2 + 121666 * (4 * xt * zt))
-//               = p * (d + 121666 * p)
-
-        mul_4(zn,p,e)
-
-// Loop down as far as 3 (inclusive)
-
-        sub     i, i, #1
-        cmp     i, #3
+        ldr     x0, [swap]
+        cmp     x0, x2
+        str     x2, [swap]
+
+// The following inner loop code is derived closely following Lenngren's
+// implementation available at "https://github.com/Emill/X25519-AArch64".
+// In particular, the basic dataflow and the organization between integer
+// and SIMD units is identical, with only a few minor changes to some
+// individual instructions (for miscellaneous reasons). The scheduling
+// was redone from scratch by SLOTHY starting from Hanno Becker's
```

Comment:
Can we say "... starting from the un-interleaved developed by Becker et. al in the SLOTHY paper", or just "... developed in the SLOTHY paper [ref]"?

Reply:
Thank you June for the additional followup work and the careful review!

I will fix those attributions as Hanno suggests now before we merge.